AP_Periph: added FreeflyRTK peripheral support

tridge opened the pull request:
This is a F732 based RTK GPS using F9P

magicrub commented on Tools/AP_Periph/rc_out.cpp:
```diff
@@ -24,20 +24,18 @@
 #define SERVO_OUT_RCIN_MAX      16  // SRV_Channel::k_rcin1 ... SRV_Channel::k_rcin16
 #define SERVO_OUT_MOTOR_MAX     12  // SRV_Channel::k_motor1 ... SRV_Channel::k_motor8, SRV_Channel::k_motor9 ... SRV_Channel::k_motor12
 
-#if HAL_PWM_COUNT == 0
```
What is this accomplishing in this particular PR? What is this new board using from the servo library? SRV_Channel.h has this logic which will cause it to have 16 servos and thus a bunch of params that do nothing..?

```
#if !defined(NUM_SERVO_CHANNELS) && defined(HAL_BUILD_AP_PERIPH) && defined(HAL_PWM_COUNT) && (HAL_PWM_COUNT >= 1)
    #define NUM_SERVO_CHANNELS HAL_PWM_COUNT
#else
    #define NUM_SERVO_CHANNELS 16
#endif
```

The reason I put in the HAL_PWM_COUNT==0 check is if NUM_SERVO_CHANNELS=0 then all hell breaks loose throughout all other libraries that use NUM_SERVO_CHANNELS and lots of changes are need to wrap them to allow=0. Did you run into any of that? Maybe that's been handled since then?

tridge replied:
its needed for notify for LEDs

magicrub reviewed (commented):
I'd *really* like to convert all our hwdefs to use a hwdef-common.dat format to simply the design. It would be nice if we convert them all over but I just don't see that happening in the short-term due to having to confirm nothing breaks on every platform. Meanwhile, it would be nice to set a good example. I've done that with 2 branches so far but neither are in master yet so I won't enforce it on PRs quire yet...

btw, I have a feeling you added RC_OUT just to add the safety_state for the button. If so, I've already implemented a better fix to move it out of rc_out. See https://github.com/maviday/ardupilot/blob/periph_ESC/Tools/AP_Periph/can.cpp#L523-L527

peterbarker commented:
On Fri, 16 Apr 2021, Tom Pittenger wrote:

> I'd really like to convert all our hwdefs to use a hwdef-common.dat format to simply the design. It would be nice if we convert them all over but I just don't see that happening in the short-term due to having to confirm nothing breaks on every platform. Meanwhile, it would be nice to set a good example. I've

That's been proven before by making sure that the generated hwdef.h files 
are identical IIRC.

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/FreeflyRTK/hwdef.dat:
```diff
@@ -0,0 +1,159 @@
+
+# store parameters in pages 2 and 3
+define STORAGE_FLASH_PAGE 2
+define HAL_STORAGE_SIZE 8192
```
its got 512k of flash, and plenty of room for params, why not use it?

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/common/flash.c:
```diff
@@ -111,6 +111,10 @@ static const uint32_t flash_memmap[STM32_FLASH_NPAGES] = { KB(32), KB(32), KB(32
 #define STM32_FLASH_NPAGES  12
 static const uint32_t flash_memmap[STM32_FLASH_NPAGES] = { KB(32), KB(32), KB(32), KB(32), KB(128), KB(256), KB(256), KB(256),
                                                            KB(256), KB(256), KB(256), KB(256) };
+#elif BOARD_FLASH_SIZE == 512
```
fixed

tridge commented:
> I'd _really_ like to convert all our hwdefs to use a hwdef-common.dat format to simply the design.

I'd prefer to fix chibios_hwdef.py to automatically fill in good defaults for AP_Periph when env AP_PERIPH is set to 1. So we need less fluff in there in the first place

tridge commented on libraries/SRV_Channel/SRV_Channel_aux.cpp:
```diff
@@ -72,19 +72,23 @@ void SRV_Channel::output_ch(void)
  */
 void SRV_Channels::output_ch_all(void)
 {
+#if NUM_SERVO_CHANNELS
```
from dev call, preference is for a FOR_EACH_SERVO_CHANNEL macro

tridge replied:
peter is happy with pragam for suppressing error instead

eovnu87435ds commented:
I've been doing some hardware testing, so I'd like to report my progress thus far.

**Tests:**

- [x] Bootloader and applications running
- [x] GNSS data from air and ground modules
- [x] Mag data 
- [x] Ground survey in
- [x] Static CAN node assignment
- [x] GNSS over serial
- [x] NTF LED over serial
- [ ] Firmware upload over USB
- [ ] Orientation of compass
- [ ] Compass calibration success
- [ ] RTK for 1+ hour over radio link
- [ ] Test Flight

**Confirmed to be NOT working:**

- No RGB LED over CAN. Confirmed to not be a power issue. The LED works correctly with a direct I2C connection.


- No Safety Switch. The switch is connected to PC13 and reads 3v3 at the MCU when unpressed and 0v when pressed. replacing this unit with an mRo kitcan on the bus gives a working safety switch.

- No Baro readings are found in UAVCAN GUI.

tridge commented:
> I've been doing some hardware testing, so I'd like to report my progress thus far.

thanks!
I've fixed the LED and baro, and fixed the mag orientation.
Still puzzled by safety switch

tridge commented:
> 

this is working for me with uploader.py, but not with MissionPlanner

tridge commented:
I've now got safety LED and switch working

eovnu87435ds commented:
Flight Tested with newest fixes (safety switch, baro, LED)!

[00000199.zip](https://github.com/ArduPilot/ardupilot/files/6444499/00000199.zip)

I haven't yet re-tested the other aspects for any regressions.

tridge commented:
> Flight Tested with newest fixes (safety switch, baro, LED)!

looks good!
![image](https://user-images.githubusercontent.com/831867/117517558-af62fd00-afdf-11eb-8ccc-94f98776eeb1.png)
![image](https://user-images.githubusercontent.com/831867/117517566-b68a0b00-afdf-11eb-8b52-fe08e5006ce5.png)
![image](https://user-images.githubusercontent.com/831867/117517586-c4d82700-afdf-11eb-939b-26efc2466d5a.png)

tridge commented:
> Do we need the .ioc file (mxcube ?)

they are very useful for making clock tree changes, and for working out things like I2C timings. I quite often refer users to them.

tridge commented:
this is ready to go, just needs another dev to approve

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32F732xx.py:
```diff
@@ -0,0 +1,371 @@
+    # flags of 1 means DMA-capable
+    # flags of 2 means faster memory for CPU intensive work
+    'RAM_MAP' : [
+        (0x20000000,  64,  3), # DTCM
```
You have a comment above saying DTCM DMA does not work in which case this option is wrong

tridge replied:
where is that comment? DMA is fine on DTCM for F732.

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/common/bouncebuffer.c:
```diff
@@ -24,6 +24,9 @@
 // always use a bouncebuffer on H7, to ensure alignment and padding
 #define IS_DMA_SAFE(addr) false
+#elif defined(STM32F732xx)
+// DTCM DMA does not work on F732
```
Here is the comment

tridge replied:
thanks, I've fixed the comment. I didn't change the implementation as I've already spent a lot of time testing and I'm happy with the results
